src/message_filters/__init__.py

```diff
@@ -227,27 +238,32 @@ class ApproximateTimeSynchronizer(TimeSynchronizer):
     """
     Approximately synchronizes messages by their timestamps.
 
-    :class:`ApproximateTimeSynchronizer` synchronizes incoming message filters by the
-    timestamps contained in their messages' headers. The API is the same as TimeSynchronizer
-    except for an extra `slop` parameter in the constructor that defines the delay (in seconds)
-    with which messages can be synchronized. The ``allow_headerless`` option specifies whether
-    to allow storing headerless messages with current ROS time instead of timestamp. You should
+    :class:`ApproximateTimeSynchronizer` synchronizes incoming message filters
+    by the timestamps contained in their messages' headers. The API is the same
+    as TimeSynchronizer except for an extra `slop` parameter in the constructor
+    that defines the delay (in seconds) with which messages can be synchronized.
+    The ``allow_headerless`` option specifies whether to allow storing
+    headerless messages with current ROS time instead of timestamp. You should
     avoid this as much as you can, since the delays are unpredictable.
     """
 
     def __init__(self, fs, queue_size, slop, allow_headerless=False):
         TimeSynchronizer.__init__(self, fs, queue_size)
-        self.slop = rospy.Duration.from_sec(slop)
+        self.slop = Duration(seconds=slop).nanoseconds
         self.allow_headerless = allow_headerless
 
     def add(self, msg, my_queue, my_queue_index=None):
         if not hasattr(msg, 'header') or not hasattr(msg.header, 'stamp'):
             if not self.allow_headerless:
-                rospy.logwarn("Cannot use message filters with non-stamped messages. "
-                              "Use the 'allow_headerless' constructor option to "
-                              "auto-assign ROS time to headerless messages.")
+                rclpy.logging._root_logger.log("can not use message filters"
+                              "for messages without timestamp infomation when"
+                              "'allow_headerless' is disabled. auto assign"
+                              "ROSTIME to headerless messages once enabling"
+                              "constructor option of 'allow_headerless'.",
+                              LoggingSeverity.INFO)
                 return
-            stamp = rospy.Time.now()
+
+            stamp = ROSClock().now().nanoseconds
```

Comment:
Same here, use the semantic datatypes.

Reply:
So this should store the time object. An example of how this might be abused would be trying to add two times which is not valid, but it is valid for durations.

Reply:
Yes, it makes sense! But it will be problematic for the following (a.k.a line 255) case which can't be iterated and subscript for the semantic datatypes, anyway, let me try to optimize it ASAP. thanks.
```
 my_queue[stamp] = msg
```